src/aspire/operators/diag_matrix.py

```diff
@@ -0,0 +1,622 @@
+                f"DiagMatrix.matmul of ndarray only supports 2D, received {lhs.shape}."
+            )
+
+        # Then the mathematical `AD` for diag D and matrix A,
+        # is expressed in code as `A@D`.
+        #
+        # The operation `AD` is known as Right Scaling,
+        # and means to scale the column A_j by d_j in D.
+        # This can be accomplished by broadcasting an elemental multiply.
+        # In this case A is the variable `lhs`
+
+        return lhs * self._data[np.newaxis]
+
+    def mul(self, other):
+        """
+        Compute the elementwise multiplication of a `DiagMatrix` instance and a
+        scalar or another `DiagMatrix`.
+
+        :param other: The rhs in the multiplication..
+        :return: A `self` * `other` as type of `other`.
+        """
+        if isinstance(other, DiagMatrix):
+            res = DiagMatrix(self._data * other._data)
+        elif isinstance(other, np.ndarray):
+            res = self * DiagMatrix(other)
+        elif isinstance(other, BlkDiagMatrix):
+            if self.stack_shape != ():
+                raise RuntimeError(
+                    f"Mixed `mul` only implemented for singletons at this time, received {self.stack_shape}."
+                )
+            res = self.as_blk_diag(other.partition) * other
+        else:  # scalar
+            res = DiagMatrix(self._data * other)
+
+        return res
+
+    def __mul__(self, val):
+        """
+        Operator overload for `DiagMatrix` scalar multiply.
+        """
+
+        return self.mul(val)
+
+    def __rmul__(self, other):
+        """
+        Convenience function, elementwise rmul commutes to mul.
+        """
+        # `mul` will attempt all compatible multiplications.
+        return self.mul(other)
+
+    def neg(self):
+        """
+        Compute the unary negation of `DiagMatrix` instance.
+
+        :return: A `DiagMatrix` like self.
+        """
+        return DiagMatrix(-self._data)
+
+    def __neg__(self):
+        """
+        Operator overload for unary negation of `DiagMatrix` instance.
+        """
+
+        return self.neg()
+
+    def abs(self):
+        """
+        Compute the elementwise absolute value of `DiagMatrix` instance.
+
+        :return: A `DiagMatrix` like self.
+        """
+
+        return DiagMatrix(np.abs(self._data))
+
+    def __abs__(self):
+        """
+        Operator overload for absolute value of `DiagMatrix` instance.
+        """
+
+        return self.abs()
+
+    def pow(self, val):
+        """
+        Compute the elementwise power of `DiagMatrix` instance.
+
+        :param val: Value to exponentiate by.
+        :return: A `DiagMatrix` like self.
+        """
+
+        return DiagMatrix(self._data**val)
+
+    def __pow__(self, val):
+        """
+        Operator overload pow of `DiagMatrix` instance.
+        """
+
+        return self.pow(val)
+
+    @property
+    def norm(self):
+        """
+        Compute the norm of a `DiagMatrix` instance.
```

Comment:
Specify that this is the 2-norm (as opposed to 1-norm or Frobenius or …).

Reply:
Sure! Changed.